README.md

```diff
@@ -11,130 +12,128 @@ The framework will help you to:
-├── mod/
-├── lib/
-│   └── ops -> ../mod/operator/ops
-├── src/
-│   └── charm.py
-└── hooks/
-    ├── install -> ../src/charm.py
-    └── start -> ../src/charm.py  # for k8s charms per below
-```
-
-The `mod/` directory should contain the operator framework dependency as a git
-submodule:
+The dependencies of the operator framework are kept as minimal as possible;
+currently that's Python 3.5 or greater, and `PyYAML` (both are included by
+default in Ubuntu's cloud images from 16.04 on).
 
-```
-git submodule add https://github.com/canonical/operator mod/operator
-```
+If you're new to the world of juju and charms, you should probably dive into our
+[tutorial](/TBD).
 
-Then symlink from the git submodule for the operator framework into the `lib/`
-directory of your charm so it can be imported at run time:
+If you know about juju, and have written charms that didn't use the operator
+framework (be it with reactive or without), we have an [introduction to the
+operator framework](/TBD) just for you.
 
-```
-ln -s ../mod/operator/ops lib/ops
-```
+If you've gone through the above already and just want a refresher, or are
+really impatient and need to dive in, feel free to carry on down.
 
-Other dependencies included as git submodules can be added in the `mod/`
-directory and symlinked into `lib/` as well.
+## A Quick Introduction
 
-You can sync subsequent changes from the framework and other submodule
-dependencies by running:
+Operator framework charms are just Python code. The entry point to your charm is
+a particular Python file. It could be anything that makes sense to your project,
+but let's assume this is `src/charm.py`. This file must be executable (and it
+must have the appropriate shebang line).
 
-```
-git submodule update
-```
+This file must be symlinked from `dispatch` (as of juju 3.8; otherwise from
+`hooks/install`, and also from `hooks/start` if it's for k8s).
 
-Those cloning and checking out the source for your charm for the first time
-will need to run:
+You also need the usual `metadata.yaml` and `config.yaml` files, and any Python
+dependencies (maybe using some kind of virtualenv). In other words, your project
+might look like this:
 
 ```
-git submodule update --init
+.
+├── config.yaml
+├── metadata.yaml
+├── env/
+│   └── # ... your Python dependencies, including the operator framework itself
+├── src/
+│   └── charm.py
+└── dispatch -> src/charm.py
 ```
 
-Your `src/charm.py` is the entry point for your charm logic. It should be set
-to executable and use Python 3.6 or greater. At a minimum, it needs to define
-a subclass of `CharmBase` and pass that into the framework's `main` function:
+> 🛈 once `charmcraft build` works the layout will get simpler!
 
-```python
-import sys
-sys.path.append('lib')  # noqa: E402
+`src/charm.py` here is the entry point to your charm code. At a minimum, it
+needs to define a subclass of `CharmBase` and pass that into the framework's
+`main` function:
 
+```python
+#!../env/bin/python3
```

Comment:
m.... are we actually going to be putting a python3 there? If we are, where is it coming from? (if dispatch is just a symlink to charm.py there isn't a place that is building a python3.)

I'd *much* prefer not bundling a python3 executable into the charm itself.
And if the plan is to check and try to setup a virtualenv in dispatch, then dispatch can't just be a symlink to src/charm.py

Reply:
just before the example it says

> You also need the usual `metadata.yaml` and `config.yaml` files, and any Python
dependencies (maybe using some kind of virtualenv). In other words, your project
might look like this:

and it carefully avoids mentioning how that env got there.

I don't like that hashbang TBH, and maybe it would be best to leave it off entirely? The example wouldn't work without it, but for pedagogical purposes it might be better.

Reply:
If we are going to have one, I'd rather have `/usr/bin/env python3` as that is likely to actually get a python that can run. I know we don't like `sys.path.append('lib')` but it at least actually works :)

Reply:
yeah, i opted for just removing it. Putting code that actually works as intended there right now would make it messy, and distracts from what we're trying to show.
(we'll just have to make reality match this instead of the other way around)